src/broker/saslAuthenticator/scram.js

```diff
@@ -0,0 +1,320 @@
+    }
+
+    const result = []
+    for (let i = 0; i < length; i++) {
+      result.push(bufferA[i] ^ bufferB[i])
+    }
+
+    return Buffer.from(result)
+  }
+
+  /**
+   * @param {Connection} connection
+   * @param {Logger} logger
+   * @param {DigestDefinition} digestDefinition
+   */
+  constructor(connection, logger, digestDefinition) {
+    this.connection = connection
+    this.logger = logger
+    this.digestDefinition = digestDefinition
+
+    const digestType = digestDefinition.type.toUpperCase()
+    this.PREFIX = `SASL SCRAM ${digestType} authentication`
+
+    this.currentNonce = SCRAM.nonce()
+  }
+
+  async authenticate() {
+    const { PREFIX } = this
+    const { host, port } = this.connection
+    const broker = `${host}:${port}`
+
+    try {
+      this.logger.debug('Exchanging first client message', { broker })
+      const clientMessageResponse = await this.sendClientFirstMessage()
+
+      this.logger.debug('Sending final message', { broker })
+      const finalResponse = await this.sendClientFinalMessage(clientMessageResponse)
+
+      if (finalResponse.e) {
+        throw new Error(finalResponse.e)
+      }
+
+      const serverKey = await this.serverKey(clientMessageResponse)
+      const serverSignature = this.serverSignature(serverKey, clientMessageResponse)
+
+      if (finalResponse.v !== serverSignature) {
+        throw new Error('Invalid server signature in server final message')
+      }
+
+      this.logger.debug(`${PREFIX} successful`, { broker })
+    } catch (e) {
+      const error = new KafkaJSSASLAuthenticationError(`${PREFIX} failed: ${e.message}`)
+      this.logger.error(error.message, { broker })
+      throw error
+    }
+  }
+
+  /**
+   * @private
+   */
+  async sendClientFirstMessage() {
+    const clientFirstMessage = `${GS2_HEADER}${this.firstMessageBare()}`
+    const request = scram.firstMessage.request({ clientFirstMessage })
+    const response = scram.firstMessage.response
+
+    return this.connection.authenticate({
+      authExpectResponse: true,
+      request,
+      response,
+    })
+  }
+
+  /**
+   * @private
+   */
+  async sendClientFinalMessage(clientMessageResponse) {
+    const { PREFIX } = this
+    const iterations = parseInt(clientMessageResponse.i, 10)
+    const { minIterations } = this.digestDefinition
+
+    if (!clientMessageResponse.r.startsWith(this.currentNonce)) {
+      throw new KafkaJSSASLAuthenticationError(
+        `${PREFIX} failed: Invalid server nonce, it does not start with the client nonce`
+      )
+    }
+
+    if (iterations < minIterations) {
+      throw new KafkaJSSASLAuthenticationError(
+        `${PREFIX} failed: Requested iterations ${iterations} is less than the minimum ${minIterations}`
+      )
+    }
+
+    const finalMessageWithoutProof = this.finalMessageWithoutProof(clientMessageResponse)
+    const clientProof = await this.clientProof(clientMessageResponse)
+    const finalMessage = `${finalMessageWithoutProof},p=${clientProof}`
```

Comment:
I don't think naming this different will help; it's better to have a close mapping to the RFC. This is also "standardized", anyone reading SCRAM code will easily find the components of the message instead of parsing our understanding of it. WDYT?